File: PART2/ch07/service/api/src/main/java/org/delivery/api/domain/userorder/converter/UserOrderConverter.java
package org.delivery.api.domain.userorder.converter;

import org.delivery.api.common.annotation.Converter;
import org.delivery.api.domain.user.model.User;
import org.delivery.api.domain.userorder.controller.model.UserOrderResponse;
import org.delivery.db.storemenu.StoreMenuEntity;
import org.delivery.db.userorder.UserOrderEntity;

import java.math.BigDecimal;
import java.util.List;

@Converter
public class UserOrderConverter {

    public UserOrderEntity toEntity(
        User user,
        List<StoreMenuEntity> storeMenuEntityList
    ){
        var totalAmount = storeMenuEntityList.stream()
            .map(it -> it.getAmount())
            .reduce(BigDecimal.ZERO, BigDecimal::add);

        return UserOrderEntity.builder()
            .userId(user.getId())
            .amount(totalAmount)
            .build()
            ;
    }

    public UserOrderResponse toResponse(
        UserOrderEntity userOrderEntity
    ){
        return UserOrderResponse.builder()
            .id(userOrderEntity.getId())
            .status(userOrderEntity.getStatus())
            .amount(userOrderEntity.getAmount())
            .orderedAt(userOrderEntity.getOrderedAt())
            .acceptedAt(userOrderEntity.getAcceptedAt())
            .cookingStartedAt(userOrderEntity.getCookingStartedAt())
            .deliveryStartedAt(userOrderEntity.getDeliveryStartedAt())
            .receivedAt(userOrderEntity.getReceivedAt())
            .build()
            ;
    }
}

File: PART3/ch03/ch03_springcloudapigateway/ch03_03/service/api/src/main/kotlin/org/delivery/api/domain/userorder/converter/UserOrderConverter.kt
package org.delivery.api.domain.userorder.converter

import org.delivery.api.domain.user.model.User
import org.delivery.api.domain.userorder.controller.model.UserOrderResponse
import org.delivery.common.annotation.Converter
import org.delivery.db.store.StoreEntity
import org.delivery.db.storemenu.StoreMenuEntity
import org.delivery.db.userorder.UserOrderEntity

@Converter
class UserOrderConverter {

    fun toEntity(
        user: User?,
        storeEntity: StoreEntity?,
        storeMenuEntityList: List<StoreMenuEntity>?
    ): UserOrderEntity {
        val totalAmount = storeMenuEntityList
            ?.mapNotNull { it -> it.amount }
            ?.reduce { acc, bigDecimal -> acc.add(bigDecimal) }

        return UserOrderEntity(
            userId = user?.id,
            store = storeEntity,
            amount = totalAmount,
        )
    }

    fun toResponse(
        userOrderEntity: UserOrderEntity?
    ): UserOrderResponse {
        return UserOrderResponse(
            id = userOrderEntity?.id,
            status = userOrderEntity?.status,
            amount = userOrderEntity?.amount,
            orderedAt = userOrderEntity?.orderedAt,
            acceptedAt = userOrderEntity?.acceptedAt,
            cookingStartedAt = userOrderEntity?.cookingStartedAt,
            deliveryStartedAt = userOrderEntity?.deliveryStartedAt,
            receivedAt = userOrderEntity?.receivedAt
        )
    }

}